Port GRO tests to new BaseReader/Writer Test classes

utkbansal opened the pull request:
Fixes part of #516 

Changes made in this Pull Request:
 - Modified `GROWriter` to conform to the Reader API standard (works with `Universe`, `Timestep` & `AtomGroup`).
 - Breaks `BaseReaderTest` into `BaseReaderTest` and `MultiframeReaderTest` for single frame & multi frame readers.
 - All tests now inherit from Base reader/writer test classes.

PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

utkbansal commented:
@jbarnoud When I ran the `create_data.py` script, the old test files `test.xyz` and `test.trr` got changed. Is this okay? I wasn't expecting this.

kain88-de commented on testsuite/MDAnalysisTests/data/coordinates/test.xyz:
```diff
@@ -6,28 +6,28 @@ frame 0
       CA    12.00000   13.00000   14.00000
 5
-frame 0
+frame 1
```
This looks like we fixed something in the xyz writer since we created these files.

kain88-de replied:
Please also create a new zipped file for the xyz format based on this. The readme says how this is done.

kain88-de replied:
this file should be removed too

utkbansal replied:
done

kain88-de commented on testsuite/MDAnalysisTests/coordinates/test_gro2.py:
```diff
@@ -0,0 +1,24 @@
+import numpy as np
+
+
+class GROReference(BaseReference):
```
they should be included in the main gro test file.

utkbansal replied:
Yes, I just wanted to keep it separate while I'm working on it so it don't get it mixed. Once I'm done I'll move it to its correct place.

kain88-de commented on testsuite/MDAnalysisTests/data/coordinates/test.gro:
```diff
@@ -0,0 +1,8 @@
+    3ILE     CA    3   9.600  11.200  12.800  0.9600  1.1200  1.2800
+    4LYS     CA    4  14.400  16.000  17.600  1.4400  1.6000  1.7600
+    5LEU     CA    5  19.200  20.800  22.400  1.9200  2.0800  2.2400
+   8.51000   8.59223   8.35003   0.00000   0.00000   0.69131   0.00000   1.45589   2.09054
```
This looks like it's only the last frame. But according to the docs GRO can only write one frame per file. Multiple frames are stored in multiple files. So the trajectory in gro files can only be read with the chain-reader (that works and it tested separately no need to test the chain reader in this PR)

kain88-de commented on testsuite/MDAnalysisTests/data/coordinates/test.gro:
```diff
@@ -0,0 +1,8 @@
+Written by MDAnalysis
```
This header line should include a time `t = ` as mentioned in the gromacs docs.

richardjgowers replied:
This will be an issue with the GROWriter then?

utkbansal replied:
`t` is optional according to docs

kain88-de replied:
Yes but we have that information so we should provide it.

kain88-de commented on testsuite/MDAnalysisTests/coordinates/test_gro2.py:
```diff
@@ -0,0 +1,24 @@
+        self.ext = 'gro'
+        # self.volume = 0
+        # self.dimensions = np.zeros(6)
+        # self.container_format = True
```
You need to deactivate all tests that skip frames or otherwise work on the assumption that a gro stores a trajectory instead of a single snapshot. For this you need to change the BaseReaderTest.

utkbansal replied:
I looked into how tests are skipped and I came across the `skipif` decorator that we import from `numpy`.  But then to check the `ext` attribute I'd have to use `self` which I can't. I mean I can't do something like:

`@dec.skipif(self.ext == 'gro')`

Is there a workaround for this? Or should I fall back to vanilla if/else statements in methods,

kain88-de replied:
General comment. The general purpose BaseReaderTest shouldn't know anything about the specific extensions used. You should enable/disable tests based on values in the reference. 

Why doesn't the skipif work?

utkbansal replied:
@kain88-de It doesn't work because there isn't any reference to `self`.

richardjgowers replied:
It's probably best to make `BaseReaderTest` a subclass of a base class for single frame readers?  The TRR readers etc do everything that GROReader does, with some extra features (traj iteration)

utkbansal replied:
So the base class has what a GROReader should have. BaseReaderTest should then inherit from it and add functionality for `TRR` etc. that have multiple frames?

richardjgowers replied:
@utkbansal yeah so something like

```python
class BaseReaderTest(object):
    # define what ALL readers should do

class MultiframeReaderTest(BaseReaderTest):
   # define what extra a multiframe reader does

```

So the current existing `BaseReaderTest` is actually a mix of two sets of tests imo

jbarnoud commented:
I do not see why the frames were numbered from 0, and are now numbered from 1. After a quick glance at the code, I would have expected them to be numbered from 0; yet, I am not super up to date with the readers. Maybe @kain88-de has an idea?

kain88-de commented on testsuite/MDAnalysisTests/data/coordinates/test.gro:
```diff
@@ -0,0 +1,8 @@
+Written by MDAnalysis
+    5
+    1MET     CA    1   0.000   1.600   3.200  0.0000  0.1600  0.3200
```
The last 3 values are the velocities. You should also check that they are activated in the tests

utkbansal commented:
@kain88-de Can you have a look at the travis build errors. I don't understand the errors. Though there are errors in the code, but the traceback is different.

kain88-de commented:
lets see what happens when we restart the build

utkbansal commented:
@kain88-de Same error `TypeError: sequence item 0: expected string or Unicode, exceptions.TypeError found` Weird thing is that no part of the traceback has any mention of my code, its all from `nose`

jbarnoud commented:
The traceback is because of my nose plugin that look at open files. My last "fix" on it obviously made it more fragile. I'll deactivate it from travis.

utkbansal commented:
@jbarnoud Thanks! the flag `--no-open-files` almost fixes the bug and I get the correct tracebacks.
So back to the original question that I wanted to ask, 7 tests are currently failing which compare values with the desired values that are set in the `BaseReference`. How do I know the correct desired value to set in the `BaseReference`?

eg, `MDAnalysisTests.coordinates.test_gro2.TestGROReader.test_first_dimensions` fails with the following error -

```
Arrays are not almost equal to 6 decimals

(mismatch 100.0%)
 x: array([ 85.100006,  86.199959,  87.300041,  75.400002,  80.400002,
        85.400024], dtype=float32)
 y: array([ 81.099998,  82.199997,  83.300003,  75.      ,  80.      ,  85.      ], dtype=float32)

```

Here the value `y` was already set in the `BaseReference`, my guess is that I will have to change it to an appropriate value for the `GRO` format. How do I get the correct value?

richardjgowers commented:
Hey, this is good work. I think the problem here is that the Gro file is the last frame (it gets overwritten repeatedly until the last frame). So the reference is correct, just the write method which creates the test Gro file should only write the first frame.

utkbansal commented:
@richardjgowers Okay. So I will have to modify the `create_data.py` script to output only the first frame in case of `gro`, right?

richardjgowers commented:
Yes

On Sat, 4 Feb 2017, 11:18 a.m. Utkarsh Bansal, <notifications@github.com>
wrote:

> @richardjgowers <https://github.com/richardjgowers> Okay. So I will have
> to modify the create_data.py script to output only the first frame in
> case of gro, right?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/MDAnalysis/mdanalysis/pull/1196#issuecomment-277436464>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AI0jBw8N1jazX2z36AsO5ZYeeooktM3mks5rZF6IgaJpZM4L2B2S>
> .
>

utkbansal commented:
In the last commit I added velocities to the `GROReference`, but I have multiplied each velocity value in the `gro` file by `10` to compensate for the change in units.

kain88-de commented:
@utkbansal It looks like you need to patch the GroWriter since it isn't following the standard right now. You will also have to split the Writer tests into a BaseWriter and MultiFrameWriter tests.

utkbansal commented:
@kain88-de You mean the `GROWriter` should also have a `n_atoms` attribute?

kain88-de commented:
> You mean the GROWriter should also have a n_atoms attribute?

yes

> Also what about TestGROReader.test_iter_as_aux_lowf, will it also be a part of MultiframeReader

@jbarnoud you know most about the aux reader

jbarnoud commented:
`TestGROReader.test_iter_as_aux_lowf` tests if the aux reader manage to align data and frames when the data have a lower frequency as the trajectory. As it is about iterating over trajectory frames, it is only relevant when there are multiple frames.

utkbansal commented:
@jbarnoud @kain88-de Regarding `MDAnalysisTests.coordinates.test_gro2.TestGROReader.test_total_time` testcase, should the value of the total time come out to be `4` or will it be `0` because we have only one frame?

richardjgowers commented:
@utkbansal 0

utkbansal commented:
@kain88-de I don't see any frame skipping stuff in `BaseWriterTest` so I'm not sure what should be shifted to `MultiFrameWriter`.

kain88-de commented:
@utkbansal the writer tests often writes multi frame trajectories. The GRO format doesn't allow to write multiple frames into one file. So the Writer tests should only check that the one frame it can write is written correctly.

utkbansal commented:
@kain88-de @richardjgowers I have added all the tests required for the `GROReader` class, a couple of them are failing right now and I'll need help with them.

Regarding tests for `GROWriter`, I think we need to add some parts to the `GROWriter` class to make its working consistent with what the `XYZWriter` does. Specifically [this](https://github.com/MDAnalysis/mdanalysis/blob/develop/package/MDAnalysis/coordinates/XYZ.py#L196) part in the `XYZWriter` has no equivalent in the `GROWriter` class, due to which tests in `TestGROWriter` are failing.

utkbansal commented:
@kain88-de @richardjgowers @jbarnoud I need help with this please.

kain88-de commented on package/MDAnalysis/coordinates/base.py:
```diff
@@ -1747,7 +1747,7 @@ class SingleFrameReader(ProtoReader):
     _err = "{0} only contains a single frame"
 
-    def __init__(self, filename, convert_units=None, **kwargs):
+    def __init__(self, filename, convert_units=None, n_atoms=None, **kwargs):
```
why did you make this change?

kain88-de commented:
> Regarding tests for GROWriter, I think we need to add some parts to the GROWriter class to make its working consistent with what the XYZWriter does. Specifically this part in the XYZWriter has no equivalent in the GROWriter class, due to which tests in TestGROWriter are failing.

The GROWriter currently doesn't follow the standard API that we have. So you will need to change that one to accept and work with  a Timestep, AtomGroup, or a Universe. The `frame` argument should be removed. Please also rename `selection` to `obj` and document that it can be any of the types specified above.

Writing a Universe or AtomGroup you should be able to use the current code. For the Timestep you have to make sure that valid names and resnames exist. The standard for unknown atom name is `X` and for the resname `UNK`. We currently already do this for AtomGroups that don't provide a names or resnames.

utkbansal commented:
@kain88-de @richardjgowers I have tried to modify `GROWriters`'s API to accommodate `Timestep`,
could you please review.
2 tests cases in `TestGROWriter` currently fail and are probably due to unit conversions, I am looking into it.
I have also added bz2 compressed `GRO` file and added a reader test class for it.

utkbansal commented:
@richardjgowers Could you please review these changes?

richardjgowers commented on package/MDAnalysis/coordinates/GRO.py:
```diff
@@ -269,19 +275,35 @@ def write(self, selection, frame=None):
+        # else:
+        #     frame = u.trajectory.ts.frame
+
+        try:
```
@utkbansal `test_writer_large` has started to fail, I think this is because of this change.  If an AtomGroup is passed then you don't want to turn this into a Timestep, otherwise you lose a lot of atom name data etc

utkbansal replied:
@richardjgowers @kain88-de I'm a bit stuck at this point. When I try to check the type of the `atoms`  as follows `type(atoms) is AtomGroup`, it is False. Even tough my debugger says that it is indeed an `AtomGroup`.

richardjgowers replied:
`type(this) == that` is very literal and won't handle subclasses.  `AtomGroup` instances are subclasses of `groups.AtomGroup` as we customise the class for each Universe.  If you do `isinstance(atoms, groups.AtomGroup)` it should work.

utkbansal replied:
Yes, I got that. Thanks. But now there is some problem with `triclinic_dimensions`. Looking into it.

utkbansal replied:
@richardjgowers [This](https://github.com/MDAnalysis/mdanalysis/pull/1196/files#diff-221626a4bf4542baa2eeb062fb0d4368L369) line is causing the problem. AtomGroup doesn't have a triclinic dimensions. What would be a good way to handle this? I could wrap the if-else block in another if-else but that doesn't seem good.

richardjgowers replied:
I'd do something like:
```python
try:  # for AtomGroup/Universe
    tri_dims = obj.universe.coord.triclinic_dimensions
except AttributeError:  # for Timestep
   tri_dims = obj.triclinic_dimensions
```

richardjgowers commented on testsuite/MDAnalysisTests/coordinates/test_gro2.py:
```diff
@@ -0,0 +1,113 @@
+        if reference is None:
+            reference = GRONoConversionReference()
+        super(TestGROReaderNoConversion, self).__init__(reference)
+        self.reader = self.ref.reader(self.ref.trajectory, convert_units=False)
```
There are a few tests failing because there isn't a `lowf`/`highf` aux.  The clue to debugging it was that this was limited to the `NoConversion` set of tests.  When you replace `reader` here it doesn't have those auxiliaries.  If you go back to the reference you can see how these are normally added to `self.reader`

utkbansal commented:
@richardjgowers @kain88-de I've changed the GROWriter to follow the standard API and it now works with `Universe`, `Timestep` and `AtomGroup`. Now I just need to port the remaining test cases in the old file to the new classes. Could you please review?

Also I have a question, is there a way to create copies `AtomGroup` (Anything similar to the `.copy()` method in Timestep)?

richardjgowers commented:
@utkbansal there's no way to copy an AG, but there's not really anything I could do to an AG that would "damage" it so that I'd need a copy.  What where you thinking of needing a copy for?

kain88-de commented on package/MDAnalysis/coordinates/GRO.py:
```diff
@@ -269,19 +275,37 @@ def write(self, selection, frame=None):
+                if ts_full.n_atoms == atoms.n_atoms:
+                    selection = ts_full
+                else:
+                    selection = ts_full.copy_slice(atoms.indices)
```
why not copy the timestep from the atoms object directly? There isn't a need for this if/else clause. 

Also as a side note. `selection` is not a useful name for the variable. By convention in the code we use the name selection for variables that contain selection strings. You should rather rename it to `timestep`/`ts`

kain88-de commented on package/MDAnalysis/coordinates/GRO.py:
```diff
@@ -290,17 +314,17 @@ def write(self, selection, frame=None):
             missing_topology.append('names')
         try:
-            resnames = atoms.resnames
+            resnames = selection.resnames
```
I don't think a timestep has information about the resnames. You need to preserve the atomgroups for that.

richardjgowers replied:
I think `selection` can be either AG or TS at this point

utkbansal replied:
@kain88-de What should I name it, currently selection can be both `AtomGroup` or a `Timestep`

kain88-de replied:
You can name it `ag_or_ts`.

kain88-de commented on package/MDAnalysis/coordinates/GRO.py:
```diff
@@ -290,17 +314,17 @@ def write(self, selection, frame=None):
             missing_topology.append('resnames')
         try:
-            resids = atoms.resids
+            resids = selection.resids
```
see comment above

kain88-de commented on package/MDAnalysis/coordinates/GRO.py:
```diff
@@ -311,15 +335,17 @@ def write(self, selection, frame=None):
+            # Not inplace because AtomGroup is not a copy
+            positions = self.convert_pos_to_native(positions, inplace=False)
             if has_velocities:
                self.convert_velocities_to_native(velocities) 
```
if you don't want to modify the positions you shouldn't modify the velocities either.

kain88-de commented on package/MDAnalysis/coordinates/GRO.py:
```diff
@@ -352,21 +378,26 @@ def write(self, selection, frame=None):
+
             # Footer: box dimensions
-            if np.all(u.trajectory.ts.dimensions[3:] == [90., 90., 90.]):
+            if np.all(selection.dimensions[3:] == [90., 90., 90.]):
```
I think it's better to use `np.allclose(...)` here to avoid floating round of errors.

kain88-de replied:
`np.allclose` returns a simple bool so it can replace the `np.all` call.

kain88-de commented on testsuite/MDAnalysisTests/data/coordinates/test_incomplete_vel.gro:
```diff
@@ -0,0 +1,8 @@
+Written by MDAnalysis
+    5
+    1MET     CA    1   0.000   0.100   0.200  0.0000  0.0100  0.0200
+    2ARG     CA    2   0.300   0.400   0.500
```
@jbarnoud, @mnmelo, @orbeckst do we want to be able to read *broken* GRO files like that? I would at least want to print a warning that the file isn't formatted correctly.

jbarnoud replied:
These broken files are extremely common. They happen when you build a system by stitching pieces together. T he warning seems a good idea; for sure we should be able to read the files.

orbeckst replied:
I agree with @jbarnoud here. Although there's broken and broken: we should properly race an exception if some of the coordinates are missing (which can happen when I/O goes down in the middle of writing); you'd likely have the final box information also missing but just to be clear:

- read coordinates and velocities when available (correct format)
- read only coordinates if v are missing (but issue warning (logger/warn))
- raise exception if coordinates are missing

kain88-de replied:
@utkbansal The exceptions/warnings suggested by @orbeckst should be added as well.

utkbansal replied:
@kain88-de Aren't we already logging a warning in case of missing velocities? Have a look [here](https://github.com/utkbansal/mdanalysis/blob/728420a8db4749ee22ad05826fb9aaa9fab6a44d/package/MDAnalysis/coordinates/GRO.py#L169).

kain88-de replied:
Do we have a test for these?

utkbansal replied:
To check if the warning is raised? No.

kain88-de replied:
We should also add tests if the warning is raised and an appropriate exception text when a coordinate is missing in a row.

richardjgowers replied:
@kain88-de the issue he's trying to fix is just porting the tests to use the new fixture, which I think he's done.  He's not liable for fixing everything about the Gro tests, maybe you should raise a new Issue if this needs doing.

richardjgowers commented on package/MDAnalysis/coordinates/GRO.py:
```diff
@@ -269,19 +275,37 @@ def write(self, selection, frame=None):
+                selection = obj.copy()
+            else:
+                raise TypeError("No Timestep found in obj argument")
+        finally:
```
I think you can just get rid of this entire `finally` block.  `obj` is either - `AtomGroup`/`Universe` covered by `obj.atoms`, or `Timestep` which is covered by the `isinstance` call, else `TypeError`

richardjgowers commented on package/MDAnalysis/coordinates/GRO.py:
```diff
@@ -352,21 +378,26 @@ def write(self, selection, frame=None):
+                        pos=positions[atom_index]
                     ))
 
+            try:  # for AtomGroup/Universe
```
You can move this try/except in to the `else` block below, so it's only created when needed

utkbansal commented on package/MDAnalysis/coordinates/GRO.py:
```diff
@@ -237,26 +239,30 @@ class GROWriter(base.WriterBase):
+
         """
         self.filename = util.filename(filename, ext='gro')
+        self.n_atoms = kwargs.get('n_atoms', None)
```
@kain88-de I think we can also get rid of `**kwargs`. It is not being used anywhere in the code now.

kain88-de commented on package/MDAnalysis/coordinates/GRO.py:
```diff
@@ -239,19 +240,20 @@ class GROWriter(base.WriterBase):
     fmt['xyz_v'] = fmt['xyz'][:-1] + "{vel[0]:8.4f}{vel[1]:8.4f}{vel[2]:8.4f}\n"
 
-    def __init__(self, filename, convert_units=None,  **kwargs):
+    def __init__(self, filename, convert_units=None, n_atoms=None):
```
You need to keep the **kwargs sinkhole argument though to support our API which allows different keyword args for different formats

kain88-de commented on package/MDAnalysis/coordinates/GRO.py:
```diff
@@ -239,19 +240,20 @@ class GROWriter(base.WriterBase):
-            *filename*
-                output filename
+        Parameters
+        -----------
```
There should be a space before the double point

kain88-de commented on package/MDAnalysis/coordinates/GRO.py:
```diff
@@ -260,12 +262,10 @@ def __init__(self, filename, convert_units=None,  **kwargs):
-              optionally move to frame number *frame*
+        Parameters
+        -----------
+        obj
```
Obj : atomgroup / universe / timestep

kain88-de commented on package/MDAnalysis/coordinates/GRO.py:
```diff
@@ -237,29 +240,31 @@ class GROWriter(base.WriterBase):
-              optionally move to frame number *frame*
+        Parameters
+        -----------
+        obj : AtomGroup / Universe / Timestep
```
You still need to add new version changed strings below. Also after the first Paramter/Returns only descriptions with headings should be in the doc string

kain88-de commented on package/MDAnalysis/coordinates/GRO.py:
```diff
@@ -269,19 +274,21 @@ def write(self, selection, frame=None):
-        atoms = selection.atoms
-        # can write from selection == Universe (Issue 49)
-        coordinates = atoms.positions
+        try:
```
Actually you can probably put all the try/except blocks into one and just check if the obj isn't a Timestep. That would make the code easier to follow.

utkbansal replied:
By all try/except block you mean the ones extracting `names`, `resnames`, `resids` data?

kain88-de replied:
yes. Either they all exist or none based on the possible type of `ag_or_ts`

utkbansal replied:
Something like this?

```
        try:

            # make sure to use atoms (Issue 46)
            ag_or_ts = obj.atoms
            velocities = ag_or_ts.velocities
            names = ag_or_ts.names
            resnames = ag_or_ts.resnames

        except AttributeError:
            if isinstance(obj, base.Timestep):
                ag_or_ts = obj.copy()
                names = itertools.cycle(('X',))
                missing_topology.append('names')
                resnames = itertools.cycle(('UNK',))
                missing_topology.append('resnames')
            else:
                raise TypeError("No Timestep found in obj argument")
```

richardjgowers replied:
AtomGroup might be missing `resnames` (for example) too.  Here you just need

```python
try:
    obj = obj.atoms
except AttributeError:
    if not isinstance(obj, Timestep):
        raise TypeError
```
Which converts the three possible inputs (U, AG, TS) into the two allowed (AG or TS)

Then going forward you need to treat AG and TS the same, which is possible with the `except (AE, NDE)` blocks

utkbansal replied:
This is what happens currently.

richardjgowers replied:
Yeah this is fine as is then, I remember it being longer

richardjgowers commented:
I think we just need one try/except up top to coerce it into either AG or
TS, the rest can stay

On Tue, 14 Mar 2017, 6:38 p.m. Max Linke, <notifications@github.com> wrote:

> *@kain88-de* commented on this pull request.
> ------------------------------
>
> In package/MDAnalysis/coordinates/GRO.py
> <https://github.com/MDAnalysis/mdanalysis/pull/1196#discussion_r105993658>
> :
>
> >
> -        # make sure to use atoms (Issue 46)
> -        atoms = selection.atoms
> -        # can write from selection == Universe (Issue 49)
> -        coordinates = atoms.positions
> +        try:
>
> yes. Either they all exist or none based on the possible type of ag_or_ts
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/MDAnalysis/mdanalysis/pull/1196#discussion_r105993658>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AI0jBxP2QfJPG0qXrfK35miUaVzf5L_Mks5rlt6xgaJpZM4L2B2S>
> .
>

kain88-de commented on testsuite/MDAnalysisTests/coordinates/test_gro2.py:
```diff
@@ -0,0 +1,119 @@
+import numpy as np
```
If you think you are done these tests should be included into the main gro test file

richardjgowers replied:
This is all that needs doing before merging

kain88-de commented on testsuite/MDAnalysisTests/coordinates/test_gro2.py:
```diff
@@ -117,3 +197,82 @@ def __init__(self, reference=None):
+    name = "GRO"
+    has_box = True
+    set_box = True
+    unitcell = np.array([10., 11., 12.,
```
Are those sensible values for a triclinic box description? Have a look at the [gromacs manual](http://manual.gromacs.org/documentation/2016.3/manual-2016.3.pdf) section 3.2 for definition of periodic boundary boxes in molecular dynamics simulations.

utkbansal replied:
I have just moved this class as-is from the old `test_gro.py`. I'll have a look.

utkbansal replied:
@kain88-de Most of what was written there (except the equations) didn't make sense to me. 😞

utkbansal commented:
@richardjgowers @kain88-de I am wrapping up this PR, just a couple of things left. There are some test cases in `TestGROReader` & `TestGROReaderNoConversion` in `test_gro.py` that haven't been ported yet to the new code. They mostly include comparisons with values in `RefAdK`. How should I port them? They use the `GRO`(data/adk_oplsaa.gro) file to create the universe and the new tests use the `test.gro` file for the same so I can't straightaway use the values in `RefAdK`

Other than this all other test cases have been ported. Once the above mentioned test cases are ported too I can delete the old `test_gro.py` and rename `test_gro2.py` to `test_gro.py` and we can merge.

utkbansal commented:
@richardjgowers @kain88-de review please

kain88-de commented:
@utkbansal I'm very busy right now. I want get to this until next weekend. Maybe @richardjgowers can look into merging the two test files. Our general approach is that we try not to delete old tests.

richardjgowers commented:
Ok I'm going to give this a final review tomorrow, I want to use my proper dev machine to check a few things

richardjgowers commented:
@utkbansal looks fine to me, we can keep the old tests for now.

@kain88-de we can open a new issue about updating the xyz file if you want to get that done.  @utkbansal maybe don't commit the change to the xyz file so the compressed and uncompressed versions are identical for now.

kain88-de commented:
the test.trr update should also be removed from this PR.

utkbansal commented:
@kain88-de @richardjgowers I've reverted the changes to `test.trr`

utkbansal commented:
@richardjgowers Can this be merged now? It's complete from my side.

kain88-de commented:
@utkbansal there are merge conflicts. Please **rebase** against develop and solve these conflicts.

utkbansal commented:
@kain88-de @richardjgowers I have done a rebase and force push.

richardjgowers commented:
LGTM assuming tests pass

kain88-de commented:
@utkbansal thanks for doing this all the way to the end

utkbansal commented:
I should be the one thanking you all 😄 Without help from all of you I wouldn't be able to do this.
